gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java

```diff
@@ -178,6 +178,10 @@ public static ClientContext create(StubSettings settings) throws IOException {
     // A valid EndpointContext should have been created in the StubSettings
     EndpointContext endpointContext = settings.getEndpointContext();
     String endpoint = endpointContext.resolvedEndpoint();
+    String mtlsEndpoint = "";
+    mtlsEndpoint = settings.getMtlsEndpoint();
+    String endpointOverride = "";
+    endpointOverride = settings.getEndpointContext().transportChannelProviderEndpoint();
```

Comment:
@zhumin8 , @lqiu96 I would like the value of `endpointOverride` to be the endpoint override specified by the client. E.g. how it's done in Go: https://github.com/googleapis/google-cloud-go/blob/main/auth/internal/transport/s2a.go#L117 . Can you confirm that the `transportChannelProviderEndpoint` represents the client specified endpoint override? I also see `clientSettingsEndpoint`, but there is also logic:

```
// TransportChannelProvider's endpoint will override the ClientSettings' endpoint
      String customEndpoint =
          transportChannelProviderEndpoint() == null
              ? clientSettingsEndpoint()
              : transportChannelProviderEndpoint();
```

which shows `transportChannelProviderEndpoint` overrides `clientSettingsEndpoint`

Reply:
Users have two ways of setting a custom endpoint with Java client libraries. transportChannelProviderEndpoint will be the final endpoint if set. The logic is as follows:
1. Use transportChannelProviderEndpoint if set
2. Otherwise, use clientSettingsEndpoint if set
3. Otherwise, construct a default endpoint for the user.

Reply:
Can you explain a bit on what is the intension of `endpointOverride`? Especially, I wonder how it is different from `String endpoint = endpointContext.resolvedEndpoint();`, or if needed, does it make sense to add logic to identify endpointOverride inside `endpointContext`?

Reply:
Thank you for outlining how custom endpoint is set in Java. 

From the logic @lqiu96 mentioned above, I think we should be setting 
```
endpointOverride = ""
if (!settings.getEndpointContext(). transportChannelProviderEndpoint().isEmpty()) {
endpointOverride = settings.getEndpointContext(). transportChannelProviderEndpoint();
} else if (!settings.getEndpointContext().clientSettingsEndpoint().isEmpty()) {
endpointOverride = settings.getEndpointContext().clientSettingsEndpoint()
}
```

As @zhumin8 pointed out that this type of logic to set `endpointOverride` might make more sense to have in `EndpointContext.java`.

Responding to @zhumin8 comment inline below.

Can you explain a bit on what is the intension of endpointOverride? 
> The goal with `endpointOverride` is: we <strong>do not</strong> want to use our flow (mTLS via S2A) if a custom endpoint is set, the intension is for `endpointOverride` to represent this custom endpoint set by the client. If a default TLS endpoint is being used (point 3 in the above comment), not a custom endpoint, it is ok to upgrade to our flow (mTLS via S2A), in this case, we require that the `mtlsEndpoint` field be set. 

Especially, I wonder how it is different from String endpoint = endpointContext.resolvedEndpoint()
> IIUC, `endpoint` represents the resolved endpoint, so it could either be the custom endpoint, if a custom endpoint is set via `transportChannelProviderEndpoint`/`clientSettingsEndpoint`, or the default TLS endpoint. We specifically <strong>do not</strong> want to upgrade the connection to mTLS via S2A if a client specifies a custom endpoint. 

if needed, does it make sense to add logic to identify endpointOverride inside endpointContext?
> This is a good point. Happy to create a new field in `EndpointContext` called `endpointOverride` and resolve it there if it makes more sense. 


WDYT @zhumin8 and @lqiu96 ? Please let me know if I can clarify anything further.

Reply:
I would prefer to keep this logic internally inside EndpointContext and not adding a public setter to the settings class if possible. 

I think I understand the use case for endpointOverride, thanks!

Reply:
endpointOverride resolved in EndpointContext.java in https://github.com/googleapis/sdk-platform-java/pull/3326/commits/b320cc2ee8e0b1ae7a9cc711a20e95b8e827e7d4